Sources/OpenCombine/Publishers/Publishers.Map.swift

```diff
@@ -32,70 +43,121 @@ extension Publishers {
+    {
+        let inner = Inner(downstream: subscriber, transform: transform)
+        upstream.receive(subscriber: inner)
+    }
+}
+
+private class _Map<Upstream: Publisher, Downstream: Subscriber>
+    : OperatorSubscription<Downstream>,
+      CustomStringConvertible,
+      Subscription
+{
     typealias Input = Upstream.Output
-    typealias Failure = Downstream.Failure
-    typealias Transform = (Input) -> Downstream.Input
+    typealias Failure = Upstream.Failure
+    typealias GenericTransform = (Input) -> Result<Downstream.Input, Downstream.Failure>
 
-    private let _downstream: Downstream
-    private var _upstreamSubscription: Subscription?
-    private let _transform: Transform
+    private let _genericTransform: GenericTransform
 
-    init(downstream: Downstream, transform: @escaping Transform) {
-        _downstream = downstream
-        _transform = transform
+    init(downstream: Downstream, genericTransform: @escaping GenericTransform) {
+        _genericTransform = genericTransform
+        super.init(downstream: downstream)
     }
 
+    var description: String { return "Map" }
+
     func receive(subscription: Subscription) {
-        _upstreamSubscription = subscription
-        _downstream.receive(subscription: self)
+        upstreamSubscription = subscription
+        downstream.receive(subscription: self)
     }
 
     func receive(_ input: Input) -> Subscribers.Demand {
-        return _downstream.receive(_transform(input))
-    }
-
-    func receive(completion: Subscribers.Completion<Failure>) {
-        _downstream.receive(completion: completion)
+        switch _genericTransform(input) {
+        case .success(let output):
+            return downstream.receive(output)
+        case .failure(let error):
+            downstream.receive(completion: .failure(error))
+            return .none
+        }
     }
 
     func request(_ demand: Subscribers.Demand) {
-        _upstreamSubscription?.request(demand)
+        upstreamSubscription?.request(demand)
     }
 
-    func cancel() {
-        _upstreamSubscription?.cancel()
-        _upstreamSubscription = nil
+    override func cancel() {
+        upstreamSubscription?.cancel()
     }
 }
 
-// MARK: - CustomStringConvertible
-extension Inner: CustomStringConvertible {
-    public var description: String { return "Map" }
+extension Publishers.Map {
+
+    private final class Inner<Downstream: Subscriber>
+        : _Map<Upstream, Downstream>,
+          Subscriber
+        where Downstream.Failure == Upstream.Failure
+    {
+        init(downstream: Downstream,
+             transform: @escaping (Upstream.Output) -> Downstream.Input) {
+            super.init(downstream: downstream,
+                       genericTransform: { input in .success(transform(input)) })
+        }
+
+        func receive(completion: Subscribers.Completion<Failure>) {
+            downstream.receive(completion: completion)
```

Comment:
If `OperatorSubscription` conformed to subscription, we would not have to have this receive here in the Map.Inner.